### DISCUSSION TITLE
Can Neper be used to generate a magnetic microstructure based on anisotropy or coercivity? Also, can it simulate diffusion?

### DISCUSSION THREAD

**Discussion by mikekaylor98** [neperfepx/neper#409] [Q&A] [answered]
I want to know if this is possible and how I would even go about doing it. I also am confused by this figure in a paper I am reading, I want to recreate it. I want to be able to get data like grain misorientation angle and such, and I want to know how I could generate a microstructure that has diffusion.
<img width="324" alt="diffusion" src="https://user-images.githubusercontent.com/42279419/158083123-8684d037-5331-492f-bc02-9d8d5cac3aac.png">
This is a microstructure the paper created, Neper is the tool they used for the other figures, and I believe for this Dy diffusion as well, but I'm not entirely sure. If they did use Neper I'm not sure how to even begin making a microstructure that looks close to this. I know how to color something by z position or by any other property, but not how to color a grain differently if it is on the edge or surface of the structure. I also don't know if I can actually define a microstructure based on compound properties or not, or if I can do two different compounds in the same structure.

**Comment by rquey** [comment, marked as answer]
This may be indicated in the paper, but the microstructure was probably generated using `-morpho gg` (grain-growth polycrystal):

```
neper -T -n 1000 -morpho gg -o gg
```

You can determine if a grain is on a domain face or edge from its `xmin`, `xmax`, `ymin`, `ymax`, `zmin` and `zmax`. Such grains can define groups using option [-group](https://neper.info/doc/neper_t.html#cmdoption-group):

- Faces (group 1 = surface, group 2 = others):

```
# define groups
neper -T -loadtess gg.tess -group "(zmin<0.001||zmax>0.999)?1:2"

# visualize (red/green)
neper -V gg.tess -datacellcol group -print img1

# visualize (gray/light gray)
neper -T -loadtess gg.tess -statcell group
awk '{if ($1==1) {print "gray"} else {print "lightgray"}}' gg.stcell > col
neper -V gg.tess -datacellcol "col:file(col)" -print img2
```
![image](https://user-images.githubusercontent.com/2659376/158148768-d359c0d6-6fa6-4633-9c4b-f533dc90398f.png)
![image](https://user-images.githubusercontent.com/2659376/158148810-b5bdf357-7020-427f-8fee-6a958b4fc5c4.png)

- Edges (group 1 = edges, group 2 = others:

```
# define groups
neper -T -loadtess gg.tess -group "((zmin<0.001||zmax>0.999)&&(xmin<0.001||xmax>0.999||ymin<0.001||ymax>0.999))?1:2"

# visualize (red/green)
neper -V gg.tess -datacellcol group -print img3

# visualize (gray/light gray)
neper -T -loadtess gg.tess -statcell group
awk '{if ($1==1) {print "gray"} else {print "lightgray"}}' gg.stcell > col
neper -V gg.tess -datacellcol "col:file(col)" -print img4
```

![image](https://user-images.githubusercontent.com/2659376/158149282-0ef7cf31-cdd8-48c2-9c61-6796a0456c89.png)
![image](https://user-images.githubusercontent.com/2659376/158149300-fa3763bd-47e8-47a6-a3ed-531edf6ca52b.png)

You can even do all in one:
```
neper -T -loadtess gg.tess -group "(zmin<0.001||zmax>0.999)&&(xmin<0.001||xmax>0.999||ymin<0.001||ymax>0.999)?1:((zmin<0.001||zmax>0.999)?2:3)"
```

And Neper does not do diffusion (or any other simulation) by itself.

**Reply by mikekaylor98** [reply]
Thank you so much for your answer, this is extremely useful to me. The generation parameters do appear to be a little more complex that I thought they could be. Based on your response and the lack of mention on the site itself, I will assume that Neper is only used to generate synthetic microstructures. It is not a tool to generate a structure based on data from an actual chemical compound, just based on it's own algorithms.

**Reply by rquey** [reply]
> It is not a tool to generate a structure based on data from an actual chemical compound, just based on it's own algorithms.

Correct. Neper is math / geometry / meshing, and there is no physics involved in the tessellation generation.

**Comment by Mulike-568** [comment]
Hello,
Sir,could you please share the title of this paper with me?

**Reply by mikekaylor98** [reply]
Unfortunately, I do not have access to my data currently. Luckily, I was able to use some search terms. I will list two related (if not the same) papers.

Magnetic microstructure machine learning analysis:
https://iopscience.iop.org/article/10.1088/2515-7639/aaf26d
https://arxiv.org/pdf/1808.03794.pdf

Hope that helps!
